src/scanner.c

```diff
@@ -3091,8 +3110,7 @@ static Int InitKernel (
     Int                 i;
 
     TLS(Input) = TLS(InputFiles);
-    TLS(Input)--;
-    (void)OpenInput(  "*stdin*"  );
+    (void)OpenInput_Internal(  "*stdin*", 1  );
```

Comment:
I'm happy to look at doing that. I agree this is equally fragile, but it's equally fragile but also valid C and doesn't spawn compiler warnings, as opposed to undefined behaviour.

I have a (slight) preference for putting this fix into 4.8, as a minimal change which deals with the warning people were getting, assuming we can convince ourselves it's "obviously" no worse than what we had before, then putting something better in master.